server/aesdsocket.c

```diff
@@ -34,29 +38,111 @@
+	bool exec_flag;
+}client_param_t;
+
+typedef struct node{
+	client_param_t thread_param;							//Stores all the necessary parameters for a client thread
+	TAILQ_ENTRY(node) nodes;								//Pointer to next node of the list
+}pnode_t;
+
+typedef TAILQ_HEAD(head_s, node) head_t;
+head_t ll_head;
+
+pnode_t* _fill_queue(head_t * head, const pthread_t thread, const int c_fd)
+{
+	struct node * e = (struct node *)malloc(sizeof(struct node));
+	if (e == NULL)
+	{
+		fprintf(stderr, "malloc failed");
+		exit(EXIT_FAILURE);
+	}
+	e->thread_param.thread_id = thread;
+	e->thread_param.client_fd = c_fd;
+	e->thread_param.exec_flag = false;
+	TAILQ_INSERT_TAIL(head, e, nodes);
+	printf("%s: client fd: %d\n", __func__, e->thread_param.client_fd);
+	e = NULL;
+	return TAILQ_LAST(head, head_s);
+}
+
+void _free_queue(head_t * head)
+{
+    struct node * e = NULL;
+    while (!TAILQ_EMPTY(head))
+    {
+        e = TAILQ_FIRST(head);
+        TAILQ_REMOVE(head, e, nodes);
+        free(e);
+        e = NULL;
+    }
+}
 
 static void sig_handler(int signo){
-	if((signo == SIGINT) || (signo == SIGTERM)){
-		if(fd_socket>0){
-			close(fd_socket);
+		syslog(LOG_DEBUG, "Caught signal,exiting");
+		pnode_t *tmp = NULL;
+		int ret = EXIT_SUCCESS, res;
+		TAILQ_FOREACH(tmp, &ll_head, nodes){
+			//Close all open client fd
+			res = shutdown(tmp->thread_param.client_fd, SHUT_RDWR);
+			if(res){
+				syslog(LOG_ERR, "shutdown failed: %s", strerror(errno));
+			}
+			printf("killing thread id: %ld", tmp->thread_param.thread_id);
+			res = pthread_kill(tmp->thread_param.thread_id, SIGKILL);
+			if(res){
+				syslog(LOG_ERR, "thread kill failed: %s", strerror(res));
+				exit(EXIT_FAILURE);
+			}
+		}
+		_free_queue(&ll_head);						//Free the entire queue
+		
+		res = pthread_kill(thread, SIGKILL);
+		if(res){
+			syslog(LOG_ERR, "cleanup thread kill failed: %s", strerror(res));
+			exit(EXIT_FAILURE);
+		}
+
+		res = pthread_kill(timethread, SIGKILL);
+		if(res){
+			syslog(LOG_ERR, "timestamp thread kill failed: %s", strerror(res));
+			exit(EXIT_FAILURE);
 		}
-		if(fd_client>0){
-			close(fd_client);
+		
+		pthread_mutex_destroy(&w_mutex);
+		if(fd_socket>0){
+			if (shutdown(fd_socket, SHUT_RDWR)){
+				ret = EXIT_FAILURE;
+				syslog(LOG_ERR, "shutdown socket: %s", strerror(errno));
+			}
 		}
+
+		// if (close(fd_socket)){
+		// 	ret = EXIT_FAILURE;
+		// 	syslog(LOG_ERR, "close socket: %s", strerror(errno));
+		// }
+
 		if(fd>0){
-			close(fd);
+			if (close(fd)){
+				ret = EXIT_FAILURE;
+				syslog(LOG_ERR, "close file: %s", strerror(errno));
+			}	
 		}
-		//deleting the file
-		int ret = remove(FILE);
-		if(ret == -1){
-			syslog(LOG_ERR, "Error removing the file");
-			closelog();
-			exit(-1);
+
+		if (unlink(FILE)){
+			ret = EXIT_FAILURE;
+			syslog(LOG_ERR, "unlink file: %s", strerror(errno));
 		}
-		syslog(LOG_DEBUG, "Caught signal,exiting");
+
 		closelog();
-	}
-	exit(0);
+
+		exit(ret);
```

Comment:
@ruchitnaik is it safe to call these many async signal unsafe functions from signal handler?
and isn't this handler too big to handle!